text/0005-syncing.md

```diff
@@ -0,0 +1,114 @@
+
+**Necessary work:**
+
+1. Start a node with the trusted hash.
+2. Ask peers for the initial block from the linear chain.
+3. Synchronize all missing dependencies.
+4. Execute deploys from the linear chain in the order of them appearing in the linear chain.
+
+`BlockExecutor` sees whether it's a `switch_block` and whether it should run an auction for era `era_id + AUCTION_DELAY` . As a result of execution, it will produce a `Block` instance (that is a block of the linear chain) but also an event `NextEraValidators(finalized_block.era_id + 1, Set<Validator>)` that informs Consensus about the next era validators. Consensus component will use new validators (and random bits from blocks between booking block and key block of the next era) and create an instance of new era.
+
+Same mechanism can be used in a node that is catching up – it will use `Block`s from the linear chain to build up its Global State via `BlockExecutor` (we need a mechanism of turning `Block` into `FinalizedBlock` as that's the input that `BlockExecutor` understands) and create new eras. 
+
+Obviously, over the course of synchronization some eras will become obsolete but the assumption is that eras are deactivated/removed by a separate mechanism. For example, `BlockExecutor` component already knows about the unbonding delay so it could send an event to a consensus component that it should deactivate an era (or if consensus component needs this knowledge it could deactivate old eras by itself).
+
+#### Variant B ####
+
+Since downloading the whole linear chain (and most importantly, the deploys) is a time-consuming operation, a faster alternative would be to just sync the Global State at the initial hash. Global State is a merkle trie so it's easily splittable and has an easy way of verifying its correctness (compare root hashes).
+
+If we choose to synchronize Global State directly though, the above won't work. We need to be able to create active eras just by looking at the Global State. This should also be fairly trivial since PoS contract can track past eras, current ones and future (auctions already executed but era not yet started). We can use information from the cache and create instances of eras that are still active.
+
+**Necessary work:**
+
+1. Start a node with the trusted hash
+2. Ask peers for the initial block from the linear chain.
+3. Synchronize Global State at the post-state hash of the initial block.
+4. Validate correctness of the Global State with the post-state hash from the initial block.
+
+
+---------------------
+Once whole linear chain is "consumed" and state is built-up, we need to synchronize DAG(s) of still active eras.
```

Comment:
By "whole linear chain", you mean the segment from the initial to the most recent block?
(And maybe back as far as the unbonding period requires?)

Reply:
Depends on what variant we go with. If it's **Variant A** then it will be the whole chain from Genesis, which is the _initial_ like you're saying although that's different from Genesis only for nodes that already have _something_ in their Linear Chain storage.

Reply:
As for **Variant B**, if we first synchronize only the Global State then we don't need the linear chain at all (of course we won't have any part of it that comes _before_ the initial hash). We will sync the Global State and then proceed to synchronize the DAG(s) – this process will create the linear chain as a side-effect of syncing (at least that was my idea for it).

Reply:
Wouldn't Variant B also cause sync of everything in everyone's privately observed DAGs, as written, possibly going back to genesis?

Reply:
Only the active, not-yet-unbonded eras' DAGs will be sync'd. That's because Consensus will create only eras that are created by the blocks in the linear chain and those are finalized blocks that should be the same for all the nodes in the network.